content/blog/example-of-mapreduce-with-ezmeral-data-fabric-mapr-database-binary-table.md

```diff
@@ -0,0 +1,365 @@
+
+
+
+    I will create the table using the hbase shell command line tool customized for EDF Database and do CRUD operations using a MapReduce application.
+
+
+
+## Prerequisite: Create a Development Environment for EDF
+
+
+
+There is already an article on the HPE Developer Portal blog that describes how to deploy a Development Environment👉: [Getting Started with Spark on MapR Sandbox][GettingStartedWithSparkOnMaprSandbox]
+
+
+However, I recommend you to read the latest official documentation first👉: [Development Environment for HPE Ezmeral Data Fabric][DevelopmentEnvironmentForHpeEzmeral]
+
+
+
+**Note**:
+Basically you can follow the instructions in the documentation, the documentation tells you to install Docker Desktop on a Mac, but you don't have to install Docker Desktop, it's fine to install the Docker Engine in a standard Linux distribution.
+
+
+**It's worth noting that installing Docker Desktop in Windows won't work.**
+
+
+I tried the following: first install WSL2 (Windows Subsystem Linux 2), then install Docker Desktop for Windows and integrate with WSL2, then run the EDF Development Environment install script, but it still fails.
+
+
+So what I ended up doing was: install VMWare on my Windows PC, create a CentOS8 VM, and run the EDF Development Environment setup script in the VM. This approach is feasible.
+Also, you can always choose the version of the Development Environment you want to deploy, you just need to change the tag of the Docker image.
+
+
+
+## MapReduce on EDF Database Binary Table
+
+
+
+EDF Database Binary Table is equivalent to the EDF version of Apache HBase, but its technical implementation is different from HBase, which is of course, because the bottom layer of EDF Database Binary Table is EDF File Store.
+
+For users, there is almost no difference between using EDF Database Binary Table and using HBase.
+
+
+Now, let's imagine that we want to build a User Notifications service.
+
+
+Since HBase does not support any operations that across rows or across tables, in order to implement operations such as Joins and Group by in RDBMS, we will use MapReduce to complete some data analysis tasks.
+
+
+
+### Create a Binary Table
+
+
+
+**Important note**:
+
+<ins>the commands in this article are all executed as the user "mapr", which is by default the admin user of the EDF Development Environment.</ins>
+
+<ins>You can also use the "root" user to create the table and run the application, but if you don't modify the ACEs of the table, the "mapr" user would be not able to see the data in the table.</ins>
+
+
+
+We are going to use the [hbase shell][HbaseShell] to create a [Binary Table][BinaryTable] inside the EDF Database.
+To be able to use the `hbase shell`, we need to install the **[mapr-hbase][MaprHbasePackage]** package first.
+
+
+
+#### For convenience of data management, I would like to create a volume for the Binary Table, the command are as following:
```

Comment:
For formatting reason, I would suggest here to NOT use the heading level 4, but rather use bold characters (enclose the sentence with **) .